In [13]:
import pandas as pd
import numpy as np
import neattext.functions as nfx

from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.pipeline import Pipeline

In [5]:
!pip install neattext

     |████████████████████████████████| 114 kB 2.6 MB/s eta 0:00:01
You should consider upgrading via the '/Users/evanarneau/Documents/EFREI/Semestre-9/CHATBOTS & RECOMMANDER SYSTEMS/Chatbot_MentalHealth/venv/bin/python3 -m pip install --upgrade pip' command.


# Data import

In [8]:
train_df['clean text'] = train_df['text'].apply(nfx.remove_userhandles)
test_df['clean text'] = test_df['text'].apply(nfx.remove_userhandles)
val_df['clean text'] = val_df['text'].apply(nfx.remove_userhandles)

In [9]:
train_df['clean text'] = train_df['clean text'].apply(nfx.remove_stopwords)
test_df['clean text'] = test_df['clean text'].apply(nfx.remove_stopwords)
val_df['clean text'] = val_df['clean text'].apply(nfx.remove_stopwords)

In [10]:
train_df

text  emotion  \
0                                i didnt feel humiliated  sadness   
1      i can go from feeling so hopeless to so damned...  sadness   
2       im grabbing a minute to post i feel greedy wrong    anger   
3      i am ever feeling nostalgic about the fireplac...     love   
4                                   i am feeling grouchy    anger   
...                                                  ...      ...   
15995  i just had a very brief time in the beanbag an...  sadness   
15996  i am now turning and i feel pathetic that i am...  sadness   
15997                     i feel strong and good overall      joy   
15998  i feel like this was such a rude comment and i...    anger   
15999  i know a lot but i feel so stupid because i ca...  sadness   

                                              clean text  
0                                  didnt feel humiliated  
1            feeling hopeless damned hopeful cares awake  
2              im grabbing minute post feel greedy wrong  
3              feeling nostalgic fireplace know property  
4                                        feeling grouchy  
...                                                  ...  
15995      brief time beanbag said anna feel like beaten  
15996  turning feel pathetic waiting tables subbing t...  
15997                           feel strong good overall  
15998                     feel like rude comment im glad  
15999                       know lot feel stupid portray  

[16000 rows x 3 columns]

# Test/Train model

In [14]:
pipe_lr = Pipeline(steps=[('cv',CountVectorizer()),('lr',LogisticRegression())])

In [17]:
x_train = train_df["clean text"]
y_train = train_df["emotion"]

pipe_lr.fit(x_train,y_train)

/Users/evanarneau/Documents/EFREI/Semestre-9/CHATBOTS & RECOMMANDER SYSTEMS/Chatbot_MentalHealth/venv/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Pipeline(steps=[('cv', CountVectorizer()), ('lr', LogisticRegression())])

In [18]:
x_test = test_df["clean text"]
y_test = test_df["emotion"]

pipe_lr.score(x_test,y_test)

0.886

The accuracy of the model is of 88.6% percent, which isn't too bad but could use some improvement with further training.

# Sample Prediction

In [20]:
sampleJ = "I am feeling super happy today"

pipe_lr.predict([sampleJ])

array(['joy'], dtype=object)

In [23]:
sampleA = "I'm pissed at my friend for betraying me like that"

pipe_lr.predict([sampleA])

array(['anger'], dtype=object)

In [24]:
sampleS = "I didnt get the gifts I wanted for christmass I am so disappointed"

pipe_lr.predict([sampleS])

array(['sadness'], dtype=object)

In [25]:
sample = "I hate that Evan disturbs me when I talk to mum"

pipe_lr.predict([sample])

array(['sadness'], dtype=object)

In [32]:
sentences = test_df['text'].head(30).tolist()

In [33]:
for sentence in sentences:
    predict = pipe_lr.predict([sentence])
    print(sentence, ":", predict)

im feeling rather rotten so im not very ambitious right now : ['sadness']
im updating my blog because i feel shitty : ['sadness']
i never make her separate from me because i don t ever want her to feel like i m ashamed with her : ['sadness']
i left with my bouquet of red and yellow tulips under my arm feeling slightly more optimistic than when i arrived : ['joy']
i was feeling a little vain when i did this one : ['sadness']
i cant walk into a shop anywhere where i do not feel uncomfortable : ['fear']
i felt anger when at the end of a telephone call : ['anger']
i explain why i clung to a relationship with a boy who was in many ways immature and uncommitted despite the excitement i should have been feeling for getting accepted into the masters program at the university of virginia : ['joy']
i like to have the same breathless feeling as a reader eager to see what will happen next : ['joy']
i jest i feel grumpy tired and pre menstrual which i probably am but then again its only been a week